### PR TITLE
[rotations] add ResourceHolderMixin to fix placement in a number of resources

### PR DESCRIPTION
**Pull request by jkhales**

### CODE REVIEW

**Comment by jkhales** [comment]
@rickwierenga this fails for `ResourceStack` curious how you want to handle that

**Comment by jkhales** [comment]
> why does it fail

my guess is that because `ResourceStack` doesn't use the automatic location calculation in `assign_child_resources` that `PlateCarrierSite` does. Wondering if this should be added to a parent class somewhere instead of playing whack a mole.

**Comment by rickwierenga** [comment]
yes, i guess in that case we should also automatically find the best location of the plate, similar to what we do with CarrierSite

File: pylabrobot/liquid_handling/liquid_handler_tests.py [hunk 1]
```diff
@@ -249,6 +254,50 @@ async def test_move_plate_onto_resource_stack_with_plate(self):
     self.assertEqual(plate2.location.z, 15)
     self.assertEqual(stack.get_absolute_size_z(), 30)
 
+  async def test_move_plate_rotation(self):
+    rotations = [0, 90, 270, 360]
+    grip_directions = [
+        (GripDirection.LEFT, GripDirection.RIGHT),
+        (GripDirection.FRONT, GripDirection.BACK),
+    ]
+    site_classes = [ResourceStack, PlateCarrierSite]
```
Comment (jkhales) [inline pylabrobot/liquid_handling/liquid_handler_tests.py:263]: anything else I should put in here?

Comment (jkhales) [reply pylabrobot/liquid_handling/liquid_handler_tests.py:263]: Yes. I like that.

Comment (jkhales) [reply pylabrobot/liquid_handling/liquid_handler_tests.py:263]: Will do it. Anything else you can think of? Tilter?

Comment (rickwierenga) [reply pylabrobot/liquid_handling/liquid_handler_tests.py:263]: tilter, heater shaker, shaker, temperature controllers, all these classes. potentially a lot easier to just create a new class SinglePlateTakingResource, override assign_child_resource there, and then have these on-deck/off-deck devices inherit from that

Comment (jkhales) [reply pylabrobot/liquid_handling/liquid_handler_tests.py:263]: Yea agreed

File: pylabrobot/plate_reading/plate_reader.py [hunk 2]
```diff
@@ -48,12 +48,16 @@ def __init__(
...
       raise ValueError("There already is a plate in the plate reader.")
     if not isinstance(resource, Plate):
       raise ValueError("The resource must be a Plate.")
-    super().assign_child_resource(resource, location=location)
+
+    if location is not None:
+      # Directly call Machine's assign_child_resource method to allow for custom location
+      Machine.assign_child_resource(self, resource, location=location, reassign=reassign)
+    else:
+      super().assign_child_resource(resource, reassign=reassign)
```
Comment (jkhales) [inline pylabrobot/plate_reading/plate_reader.py:60]: this mimics the previous behavior, but not sure that is truly the behavior we desire.

File: pylabrobot/resources/utils.py [hunk 3]
```diff
@@ -199,21 +199,3 @@ def query(
       z=z,
     ))
   return matched
-
```
Comment (jkhales) [inline pylabrobot/resources/utils.py:202]: moved to the mixin file

File: pylabrobot/resources/resource_holder.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    raise ValueError("Resource rotation must be a multiple of 90 degrees on the z axis")
+  location = {
+    0.0: Coordinate(x=0, y=0, z=0),
+    90.0: Coordinate(x=resource.get_size_y(), y=0, z=0),
+    180.0: Coordinate(x=resource.get_size_y(), y=resource.get_size_x(), z=0),
+    270.0: Coordinate(x=0, y=resource.get_size_x(), z=0),
+  }[resource.rotation.z % 360]
+  return location
+
+class ResourceHolderMixin:
```
Comment (jkhales) [inline pylabrobot/resources/resource_holder.py:23]: using a mixin avoids complexity arising from "diamond inheritance"

Comment (rickwierenga) [reply pylabrobot/resources/resource_holder.py:23]: how do we make sure `ResourceHolderMixin.assign_child_resource` has the same method header as `Resource`?

Comment (jkhales) [reply pylabrobot/resources/resource_holder.py:23]: i'm not sure i follow. isnt this the same as with a normal subclassing?

Comment (jkhales) [reply pylabrobot/resources/resource_holder.py:23]: bumping this

Comment (jkhales) [reply pylabrobot/resources/resource_holder.py:23]: any more thoughts on this @rickwierenga?

File: pylabrobot/machines/machine.py [hunk 5]
```diff
@@ -38,6 +38,7 @@ def __init__(
     backend: MachineBackend,
     category: Optional[str] = None,
     model: Optional[str] = None,
+    **_kwargs
```
Comment (rickwierenga) [inline pylabrobot/machines/machine.py:41]: kwargs seems not needed here, but maybe an artifact of using mixins?

File: pylabrobot/plate_reading/plate_reader.py (hunk 2 above)
Comment (rickwierenga) [reply pylabrobot/plate_reading/plate_reader.py:54]: I think we should implement this check in the mixin: it shouldn't update the location if one is explicitly passed

File: pylabrobot/resources/plate.py [hunk 6]
```diff
@@ -44,7 +46,7 @@ def serialize(self) -> dict:
     return {**super().serialize(), "nesting_z_height": self.nesting_z_height}
 
 
-class Plate(ItemizedResource[Well]):
+class Plate(ResourceHolderMixin, ItemizedResource[Well]):
```
Comment (rickwierenga) [inline pylabrobot/resources/plate.py:49]: why does plate need this? i thought this class was specifically for resources that hold plates

File: pylabrobot/resources/resource_holder.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  """
+  A mixin class for resources that can hold other resources, like a plate or a lid
+  """
```
Comment (rickwierenga) [inline pylabrobot/resources/resource_holder.py:26]: maybe specify that this applies translations on rotation

File: pylabrobot/resources/carrier.py [hunk 8]
```diff
@@ -241,7 +240,7 @@ def get_plate_sinking_depth(plate: Plate):
       resource.register_did_assign_resource_callback(self._update_resource_stack_location)
       self.register_did_unassign_resource_callback(self._deregister_resource_stack_callback)
 
-    return get_child_location(resource) - Coordinate(z=z_sinking_depth)
```
Comment (rickwierenga) [inline pylabrobot/resources/carrier.py:244]: _get_child_location would mean the location is explicitly specified, which should mean that the mixin doesn't apply (that's only for default values). so I think we should keep get_child_location here

File: pylabrobot/machines/machine.py (hunk 5 above)
Comment (jkhales) [reply pylabrobot/machines/machine.py:41]: When you have a bunch of subclasses with different args, they get passed up the MRO chain. This is the standard way to "ignore them"

File: pylabrobot/resources/carrier.py (hunk 8 above)
Comment (jkhales) [reply pylabrobot/resources/carrier.py:244]: seems like the opposite case in line 209?
```
    location = location or self._get_child_location(resource)
```

**Comment by jkhales** [comment]
bumping this

File: pylabrobot/machines/machine.py (hunk 5 above)
Comment (rickwierenga) [reply pylabrobot/machines/machine.py:41]: this feels like bad design. why is it necessary?

File: pylabrobot/resources/carrier.py [hunk 9]
```diff
@@ -240,8 +238,10 @@ def get_plate_sinking_depth(plate: Plate):
         z_sinking_depth = get_plate_sinking_depth(first_child)
       resource.register_did_assign_resource_callback(self._update_resource_stack_location)
       self.register_did_unassign_resource_callback(self._deregister_resource_stack_callback)
+    return -Coordinate(z=z_sinking_depth)
```
Comment (rickwierenga) [inline pylabrobot/resources/carrier.py:241]: _get_sinking_depth should not handle callbacks

Comment (rickwierenga) [reply pylabrobot/resources/carrier.py:241]: made an issue for this: https://github.com/PyLabRobot/pylabrobot/issues/246

File: pylabrobot/resources/resource_stack.py [hunk 10]
```diff
@@ -95,26 +96,42 @@ def get_actual_resource_height(resource: Resource) -> float:
       return max(get_actual_resource_height(child) for child in self.children)
     return sum(get_actual_resource_height(child) for child in self.children)
 
-  def assign_child_resource(self, resource: Resource, location: Optional[Coordinate] = None,
-    reassign: bool = False):
+
+  def get_resource_stack_size(self) -> Coordinate:
     if self.direction == "x":
       resource_location = Coordinate(self.get_size_x(), 0, 0)
```
Comment (rickwierenga) [inline pylabrobot/resources/resource_stack.py:102]: potentially just put this in get_default_child_location. It seems incorrect to have a get_size method return a coordinate

File: pylabrobot/resources/resource_stack.py [hunk 11]
```diff
@@ -95,26 +96,42 @@ def get_actual_resource_height(resource: Resource) -> float:
...
+    if len(self.children) > 0:
+      top_item = self.get_top_item()
+      if isinstance(resource, Lid) and isinstance(top_item, Plate):
+        resource_location = self.get_resource_stack_size()
+        resource_location.z -= resource.nesting_z_height
+        top_item.assign_child_resource(
+          resource,
+          location=get_child_location(resource) + resource_location
+        )
+        return
```
Comment (rickwierenga) [inline pylabrobot/resources/resource_stack.py:128]: I thought we were gonna remove this

Comment (jkhales) [reply pylabrobot/resources/resource_stack.py:128]: that's def a separate PR in my opinion. It's also a breaking change.

Comment (rickwierenga) [reply pylabrobot/resources/resource_stack.py:128]: ok

File: pylabrobot/resources/resource_stack.py (hunk 10 above)
Comment (jkhales) [reply pylabrobot/resources/resource_stack.py:102]: it's used in the assign lid to resource stack special case. I can rename to something if you don't like the `size`  word.

Comment (rickwierenga) [reply pylabrobot/resources/resource_stack.py:102]: for this particular function i don't like it :)

File: pylabrobot/resources/carrier.py (hunk 9 above)
Comment (jkhales) [reply pylabrobot/resources/carrier.py:241]: i'll add the issue to the code in a comment as well

File: pylabrobot/resources/resource_stack.py (hunk 11 above)
Comment (jkhales) [reply pylabrobot/resources/resource_stack.py:128]: https://github.com/PyLabRobot/pylabrobot/issues/247

File: pylabrobot/machines/machine.py (hunk 5 above)
Comment (jkhales) [reply pylabrobot/machines/machine.py:41]: I agree, it feels weird, but I believe it's just how python does things

Comment (jkhales) [reply pylabrobot/machines/machine.py:41]: 🤖 also agrees image

**Pull request merged**